File: 29. Implement monotonic queue for min range queries.py
"""
Task: Monotonic queue for range min queries
Time Complexity: O(n)
Space Complexity: O(k)
"""

from collections import deque

class MonotonicQueue:
    def __init__(self):
        self.deq = deque()

    def push(self, x):
        while self.deq and self.deq[-1] > x:
            self.deq.pop()
        self.deq.append(x)

    def pop(self, x):
        if self.deq and self.deq[0] == x:
            self.deq.popleft()

    def get_min(self):
        return self.deq[0]

if __name__=="__main__":
    a=[2,1,3,4,6,3,8,9,10,12,56]; k=4
    mq=MonotonicQueue()
    res=[]
    for i,v in enumerate(a):
        mq.push(v)
        if i>=k:
            mq.pop(a[i-k])
        if i>=k-1:
            res.append(mq.get_min())
    print(res)

# Task: Implement monotonic queue

File: 29. Implement monotonic queue for min range queries.cpp
/*
 Task: Monotonic queue for range minimum/maximum queries
*/

#include <bits/stdc++.h>
using namespace std;

struct MonotonicQueue
{
    deque<int> dq;

    void push(int x)
    {
        while (!dq.empty() && dq.back() > x) dq.pop_back();
        dq.push_back(x);
    }

    void pop(int x)
    {
        if (!dq.empty() && dq.front() == x)
            dq.pop_front();
    }

    int getMin()
    {
        if (dq.empty())
            throw runtime_error("empty");
        return dq.front();
    }
};

int main()
{
    vector<int> a = {2, 1, 3, 4, 6, 3, 8, 9, 10, 12, 56};
    int k = 4;
    MonotonicQueue mq;

    for (size_t i = 0; i < a.size(); ++i)
    {
        if (i < static_cast<size_t>(k) - 1)
        {
            mq.push(a[i]);
        }
        else
        {
            mq.push(a[i]);
            cout << mq.getMin() << " ";
            mq.pop(a[i - k + 1]);
        }
    }

    cout << "\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(k)
